### DISCUSSION TITLE
FDK-Cuda Iterations Leading to Increased Ram Over Time

### DISCUSSION THREAD

**Discussion by ZacharyEBirch** [astra-toolbox/astra-toolbox#452] [Q&A] [answered]
I'm having an issue with the FDK_CUDA reconstruction algorithm where I am trying to run multiple iterations and delete data structures as I go.

I am doing this in Matlab and the Astra part of the code looks like this:

    for i = 1:100
    sinogram_id = astra_mex_data3d('create', '-proj3d', proj_geom, I_log);
    rec_id = astra_mex_data3d('create', '-vol', vol_geom);
     
    cfg = astra_struct('FDK_CUDA');
    cfg.ReconstructionDataId = rec_id;
    cfg.ProjectionDataId = sinogram_id;
    
    alg_id = astra_mex_algorithm('create', cfg);
    astra_mex_algorithm('run', alg_id);
    
    rec = astra_mex_data3d('get', rec_id);
    [sinogram_id, sinogram] = astra_create_sino3d_cuda(rec, proj_geom, vol_geom);

    astra_mex_algorithm('delete', alg_id);
    astra_mex_data3d('delete', rec_id);
    astra_mex_data3d('delete', sinogram_id);
    end

Where I_log is the 3D projection data.

Each time Matlab does a pass in this for loop it will allocate some amount of memory, then free that memory. However, it doesn't free all the memory so the memory usage climbs until it panics at 100% and drops as much as it can. Sometimes this works, other times it will crash. My guess is this is an issue with Matlab's memory management but I'm not 100% certain.

![MemIssue](https://github.com/astra-toolbox/astra-toolbox/assets/107949185/f580c639-ed29-48b5-89f8-e3281230867d)

**Comment by wjp** [comment, marked as answer]
Hi,

You're not freeing the first object you assign to `sinogram_id`, because its value gets overwritten by the `astra_create_sino3d_cuda` call at the end of the loop.

**Reply by ZacharyEBirch** [reply]
Oh good catch, thanks!
